tests/integration/test_bayesian_optimization.py

```diff
@@ -142,58 +140,101 @@
+            kernel = gpflow.kernels.Matern52(variance, tf.constant([0.2, 0.2], tf.float64))
+            scale = tf.constant(1.0, dtype=tf.float64)
+            kernel.variance.prior = tfp.distributions.LogNormal(
+                tf.constant(-2.0, dtype=tf.float64), scale
+            )
+            kernel.lengthscales.prior = tfp.distributions.LogNormal(
+                tf.math.log(kernel.lengthscales), scale
             )
+            gpr = gpflow.models.GPR(
+                (data.query_points, data.observations), kernel, noise_variance=1e-5
+            )
+            gpflow.utilities.set_trainable(gpr.likelihood, False)
+            return GaussianProcessRegression(gpr)
+
+        initial_query_points = search_space.sample(5)
+        observer = mk_observer(scaled_branin if branin else simple_quadratic)
+        initial_data = observer(initial_query_points)
+        model = build_model(initial_data)
 
-            arg_min_idx = tf.squeeze(tf.argmin(dataset.observations, axis=0))
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            summary_writer = tf.summary.create_file_writer(tmpdirname)
+            with tensorboard_writer(summary_writer):
 
-            best_y = dataset.observations[arg_min_idx]
-            best_x = dataset.query_points[arg_min_idx]
+                dataset = (
+                    BayesianOptimizer(observer, search_space)
+                    .optimize(num_steps, initial_data, model, acquisition_rule)
+                    .try_get_final_dataset()
+                )
 
-            relative_minimizer_err = tf.abs((best_x - BRANIN_MINIMIZERS) / BRANIN_MINIMIZERS)
-            # these accuracies are the current best for the given number of optimization steps,
-            # which makes this is a regression test
-            assert tf.reduce_any(tf.reduce_all(relative_minimizer_err < 0.05, axis=-1), axis=0)
-            npt.assert_allclose(best_y, SCALED_BRANIN_MINIMUM, rtol=0.005)
+                arg_min_idx = tf.squeeze(tf.argmin(dataset.observations, axis=0))
+
+                best_y = dataset.observations[arg_min_idx]
+                best_x = dataset.query_points[arg_min_idx]
+
+                if branin:
+                    relative_minimizer_err = tf.abs(
+                        (best_x - BRANIN_MINIMIZERS) / BRANIN_MINIMIZERS
+                    )
+                    # these accuracies are the current best for the given number of optimization
+                    # steps, which makes this is a regression test
+                    assert tf.reduce_any(
+                        tf.reduce_all(relative_minimizer_err < 0.05, axis=-1), axis=0
+                    )
+                    npt.assert_allclose(best_y, SCALED_BRANIN_MINIMUM, rtol=0.005)
+                else:
+                    # currently assume that every rule should be able to solve this in 5 steps
+                    npt.assert_allclose(best_x, tf.constant([1.0, 1.0], tf.float64), rtol=0.05)
```

Comment:
we might need to pass tolerances as arguments, if we want to have different objectives, no?

Reply:
There are only two hard-coded objectives (controlled by the branin parameter), each with its own hard-coded tolerance. If we need more flexibility we can easily refactor.